### DISCUSSION TITLE
How to linearly improve query node query qps？

### DISCUSSION THREAD

**Discussion by neblen** [milvus-io/milvus#27629] [Q&A and General discussion]
Data size: 10 million 768 dimensions
Try: 
When segment is the default value 512MB:
1. Increase load replica 1->2->4
2. Increase the number of querynode nodes 2->4->8 
Both methods have no significant effect.

**Comment by yhmo** [comment]
Observe the CPU usage of the query nodes.
If the cpu usage is not high, two possible reasons:
1. bottleneck is in the clients
2. bottleneck is in proxy node

**Reply by neblen** [reply]
I changed the segment to 2g and reloaded the collection, but the qps still couldn’t be uploaded.

**Reply by filip-halt** [reply]
Did you drop the old collection or call compact?

**Comment by neblen** [comment]
![image](https://github.com/milvus-io/milvus/assets/49058548/3c8be87e-ae20-4038-b4ce-6b7e670b7262)
It seems that the load of proxy and querynode is not very large.
![image](https://github.com/milvus-io/milvus/assets/49058548/982b9204-208a-473e-b81b-18d0b507b2cc)
The client load is not high either.

**Reply by yhmo** [reply]
I didn't see the cpu usage of each query node.
What kind of bench tool are you using?

**Reply by neblen** [reply]
i use apache ab tool. test commands such as: ab -t 10 -k -c 30 -T application/json -p request.json "http://192.168.100.2:10554/v1/vector/search"  Can milvus query qps be measured in this way?

**Reply by yhmo** [reply]
"http://xxxx:xxxx/v1/vector/search" -------- This is the RESTful API, I think the qps is lower than RPC interfaces.

**Reply by neblen** [reply]
Normally, there is no big difference. The qps of the normal gin framework can reach 15,000.

**Reply by neblen** [reply]
If the querynode node is manually reduced, do all loaded collections need to be reloaded?

**Reply by yhmo** [reply]
No need to reload, it can automatically rebalance.

**Reply by neblen** [reply]
thanks so much~ What is the rebalance time? Recent queries will report errors: fail to get shard leaders from QueryCoord...
![image](https://github.com/milvus-io/milvus/assets/49058548/5c6695a8-a36a-4326-ba86-9fd7790e4cbf)

**Reply by yhmo** [reply]
Rebalance do these things:
- release some segments from the memory of some query nodes.
- read data from s3 to load the released segments into the memory of other query nodes.

The time cost is mainly depends on the speed of reading data from s3.

**Reply by neblen** [reply]
Will the querynode node print any logs? Use this log to confirm that rebalance is completed.

**Comment by neblen** [comment]
> Normally, there is no big difference. The qps of the normal gin framework can reach 15,000.
The current status is: the 4-node quernode qps measured by vectordbbench can reach 775, but the node measured by the ab stress testing tool is only 80. I am wondering if the qps gap will be so big after using a layer of http server?
![image](https://github.com/milvus-io/milvus/assets/49058548/3c5aa031-159b-4512-98c7-8c4693ba23ce)

**Reply by yhmo** [reply]
The vectordbbench uses multiple processes to send requests to milvus by python sdk(RPC call), to achieve high qps. I guess the bottleneck is in the httpserver of milvus.

**Comment by neblen** [comment]
![image](https://github.com/milvus-io/milvus/assets/49058548/46f59db1-4aaa-4acc-bfa3-1e749f642001)
I have a doubt: Why did I add a querynode node? This node does not seem to share the segment pressure of other nodes.

**Reply by yhmo** [reply]
The possible reason is the third query node failed to start or failed to be registered into the cluster.
If you have a monitoring service, observe the "Segment Loaded Num" metric under the Query Node section:
![Screenshot from 2023-10-13 12-03-44](https://github.com/milvus-io/milvus/assets/2282099/48eee879-3171-4e40-bd72-3e67ac85ce43)
